### PR TITLE
xGov-31 - Add arbitrary decimal precision exp log pow to AVM

### PR DESCRIPTION
**Pull request by 1m1-github**

### CODE REVIEW

**Comment by scholtz** [comment]
```Add arbitrary precision exp, log, pow to the AVM```

Maybe I am missing something, but AVM already have exp, log, pow..

https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#exp
https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#log
https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#expw

also in the text you mentioned root

https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bsqrt

If you want some opcode, i recommend starting the feature request at https://github.com/algorand/go-algorand/issues?q=is%3Aissue+pow+

The decimals operations can be achieved with the uint or byte operations there i believe..

**Comment by 1m1-github** [comment]
> Maybe I am missing something, but AVM already have exp, log, pow..

the existing opcodes are for integers. the proposed opcodes are for arbitrary precision decimals, including negatives. added 'decimal' to the title for clarity.
 
> If you want some opcode, i recommend starting the feature request at https://github.com/algorand/go-algorand/issues?q=is%3Aissue+pow+

I was told (in Discord, by John Janotti) to start with a PR to the specs repo to see what it would exactly look like...so that exists now...could make a feature request or PR to the main repo or an ARC, as ever it would move forward best.

> The decimals operations can be achieved with the uint or byte operations there i believe..

not with negative values e.g. plus the byte operations are encoding decimals as binary values (like float would), which is not ideal for many use cases, as e.g. even 0.1 cannot be represented exactly. I am proposing (see the attached PR) to use the 'decimal float' type which represents all values with perfect precision (up to available memory).

File: Proposals/xgov-31.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+title: Add arbitrary decimal precision exp, log, pow to the AVM
+author: 1m1 (1m1-github)
+company_name: 2i2i
+category: AVM
+focus_area: Defi, Science, Math
+open_source: Yes
+amount_requested: 200000
+status: Draft
```
Comment (SudoWeezy) [inline Proposals/xgov-31.md:9]: Remember to change this to "Final" if you want your proposal to go through the next voting session.

File: Proposals/xgov-31.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+title: Add arbitrary decimal precision exp, log, pow to the AVM
+author: 1m1 (1m1-github)
+company_name: 2i2i
+category: AVM
+focus_area: Defi, Science, Math
+open_source: Yes
+amount_requested: 200000
```
Comment (SudoWeezy) [inline Proposals/xgov-31.md:8]: If your project requires multiple milestones, split them into multiple proposals. See: https://github.com/algorandfoundation/ARCs/pull/209

**Comment by nonfungibleab** [comment]
Hi @1m1-github 

Would you please update your proposal based on the latest criteria for milestone-based proposals? Ping me when done, so I can add it to the forum for discussion.

Details are below, but in short, you must submit a different grant for every milestone. Submit a grant for the first milestone, if approved you get paid, and work commences. Once the milestone is complete, you submit a grant for milestone 2 with proof of milestone 1 completion. And so forth for other milestones until the project is finished.

https://arc.algorand.foundation/ARCs/arc-0034
Submit a proposal
In order to submit a proposals, a proposer needs to create a pull request on the following repository: [xGov Proposals](https://github.com/algorandfoundation/xGov).

The tutorial can be found here: [xGov proposal submission Tutorial](https://www.algorand.foundation/xgov-how-to-submit-grant-proposals).

Proposals MUST:

Follow the [template form provided](https://arc.algorand.foundation/assets/arc-0034/TemplateForm).

- Follow the rules of the xGov Proposals Repository.
- Have the status Final before the end of the temperature check.
- Milestone-based grants must submit a proposal for one milestone at a time.
- Milestones need to follow the governance periods cycle. With the current 3-months cycle, a milestone could be 3-months, 6 months, 9 months etc.
- The initial proposal must include display all milestones with clear deliverables and the amount requested must match the 1st milestone.
- Once a milestone is achieved, the team must submit a proposal for the next milestone, including ways for xGovs to verify that the milestone is completed.

**Comment by 1m1-github** [comment]
sry, forgot to ping you: @nonfungibleab

**Pull request merged**